cdap-ui/app/cdap/components/DAG/DAGProvider.tsx

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+*/
+
+import * as React from 'react';
+import { List, Map } from 'immutable';
```

Comment:
Do we know the performance implications of using List/Map from immutable vs native ones?

Reply:
Good point. Out of the top of my head I don't have it. In theory it seems it will cause delay. I am not sure how we can mitigate it but I am sure it will help us a lot in avoiding unnecessary regressions we have right now. 

The pipeline UI has become complex enough for us to force us to use Immutablejs. Weighing pros & cons with existing implementation IMHO we should use Immutablejs. For some numbers of performance https://github.com/yamalight/immutable-benchmarks